Add docs for Wait for Resource Status to Apply Configuration (#7157)
istio/istio.io: content/en/docs/reference/config/config-status/index.md

---
title: Configuration Status Field
description: Describes the role of the `status` field in configuration workflow.
weight: 21
---

{{< warning >}}
This feature is in the Alpha stage, see
[Istio Feature Status](/about/feature-stages/). Your feedback is welcome in the
[Istio User Experience discussion](https://discuss.istio.io/c/UX/23). Currently,
this feature is tested only for single, low volume clusters with a single
control plane revision.
{{< /warning >}}

Istio 1.6 and later provides information about the propagation of configuration
changes through the mesh, using the `status` field of the resource.

> 

Reviewer: ```suggestion
Status is disabled by default, and can be enabled during install with: 
{{< text bash >}}
$ istioctl manifest apply --set values.pilot.env.PILOT_ENABLE_STATUS=true --set values.global.istiod.enableAnalysis=true
{{< /text >}}
```

The `status` field contains the state of a resource's configuration with various
informational messages, including:

* The resource's readiness.
* How many data plane instances are associated with it.
* Information for the output of tools, such as `istioctl analyze`.

For example, the `kubectl wait` command monitors the `status` field to determine
whether to unblock configuration and resume. For more information, see
[Wait for Resource Status to Apply Configuration](/docs/ops/configuration/mesh/config-resource-ready/).

## View the `status` field

To view the contents of the `status` field, use the following command:

{{< text bash >}}
$ kubectl get virtualservice <service-name> -o yaml
{{< /text >}}

In the output, the `status` field contains several nested fields with details

> about the process of propagating configuration changes through the mesh. A

Reviewer 2: ```suggestion
about the process of propagating configuration changes through the mesh.
```

> child field for the condition status, also named `status`, has a value of `True`

Reviewer 2: ```suggestion
```

> about the process of propagating configuration changes through the mesh. A
> child field for the condition status, also named `status`, has a value of `True`

Reviewer: ```suggestion
about the process of propagating configuration changes through the mesh. 
```

> when Istio applied a configuration successfully.

Reviewer 2: ```suggestion
```

{{< text yaml >}}
status:
  conditions:
  - lastProbeTime: null
    lastTransitionTime: "2019-12-26T22:06:34Z"
    message: "61/122 complete"
    reason: "stillPropagating"
    status: "False"
    type: Reconciled
  - lastProbeTime: null
    lastTransitionTime: "2019-12-26T22:06:56Z"
    message: "1 Error and 1 Warning found. See validationMessages field for details"
    reason: "errorsFound"
    status: "False"
    type: PassedAnalysis
  validationMessages:
  - code: IST0101
    level: Error
    message: 'Referenced gateway not found: "bogus-gateway"'
  - code: IST0102
    level: Warn
    message: 'mTLS not enabled for virtual service'
{{< /text >}}

## The `conditions` field

> The `type` field indicates the configuration condition. It can have the

Reviewer 2: ```suggestion
Conditions represent possible states of the resource. The `type` field of a condition can have the
```

Author: Reapplied

following values:

* `PassedAnalysis`
* `Reconciled`

> When you apply a configuration, a condition of type `Reconciled` and `status`
> equals `False` is added to the `conditions` field. The status becomes `True`
> when the resource finishes reconciling. The `status` field might transition to
> `True` instantaneously, depending on the speed of the cluster.
> 
> The condition of type `PassedAnalysis` has a status of `True` or
> `False` to indicate the outcome of configuration validation. `PassedAnalysis` is
> an informational field only, meaning that it does not block the application of
> an invalid configuration. For example, it is possible for the status to indicate
> that validation failed, but applying the configuration was successful. This
> means Istio was able to set the new configuration, but the configuration was
> invalid, likely due to a syntax error or similar problem.

Reviewer 2: ```suggestion
When you apply a configuration, a condition of each of these types is added to the
`conditions` field.

The `status` field of the `Reconciled` type condition is initialized to `False`,
to indicate the resource is still in the process of being distributed to all the proxies.
When finished reconciling, the status will become `True`. The `status` field might
transition to `True` instantaneously, depending on the speed of the cluster.

The `status` field of the `PassedAnalysis` type condition will have a value of
`True` or `False` depending on whether or not istio's background analyzers have
detected a problem with your config. If `False`, the problem(s) will be detailed in the
`validationMessages` field.

The `PassedAnalysis` condition is an informational field only. It does not 
block the application of an invalid configuration. It is possible for the status to
indicate that validation failed, but applying the configuration was successful.
This means Istio was able to set the new configuration, but the configuration was
invalid, likely due to a syntax error or similar problem.
```

weight: 21
---

{{< warning >}}
This feature is in the Alpha stage, see
[Istio Feature Status](/about/feature-stages/). Your feedback is welcome in the
[Istio User Experience discussion](https://discuss.istio.io/c/UX/23). Currently,
this feature is tested only for single, low volume clusters with a single
control plane revision.
{{< /warning >}}

Istio 1.6 and later provides information about the propagation of configuration
changes through the mesh, using the `status` field in a resource's
[`CustomResourceDefinition`](https://kubernetes.io/docs/concepts/extend-kubernetes/api-extension/custom-resources/#customresourcedefinitions)
(CRD).

The `status` field contains the state of a resource's configuration with various
informational messages, including:

* The resource's readiness.
* How many data plane instances are associated with it.
* Information for the output of tools, such as `istioctl analyze`.

For example, the `kubectl wait` command monitors the `status` field to determine
whether to unblock configuration and resume. For more information, see
[Wait for Resource Status to Apply Configuration](/docs/ops/configuration/mesh/config-resource-ready/).

## View the `status` field

> To view the contents of the `status` field, use the following command:

Reviewer 2: ```suggestion
You can view the contents of the `status` field of a resource using `kubectl get`. For example, to view the
status of a virtual service, use the following command:
```

Author: Reapplied

{{< text bash >}}

> $ kubectl get virtualservice service-name

Reviewer 2: ```suggestion
$ kubectl get virtualservice <service-name>
```

Reviewer: add -o yaml, or all you will see is the resource name

Author: fixed

{{< /text >}}

> In the output, the `status` field contains several child fields with details

Reviewer 2: ```suggestion
In the output, the `status` field contains several nested fields with details
```

about the process of propagating configuration changes through the mesh. A

> child field for the condition status, also named `status`, has a value of `True`

Reviewer 2: Here it should say describe the `conditions` field first. What is that, and why are there many? One for each what?

Reviewer: "Conditions represent possible states of the resource.  In Istio 1.6, we are introducing two condition types, Reconciled and PassedAnalysis.  When the condition of type Reconciled has a status of false, that means that this resource is still in the progress of being distributed to all proxies.  When the condition of type PassedAnalysis has a status of false, this means that istio's background analyzers have detected a problem with your config, which will be detailed in the `validationMessages` field.

Reviewer: On second thought, perhaps just remove the references to True and False up here, and leave the description of the Condition field.  You explain true and false very thoroughly later on, I think.

Author: fixed

Reviewer 2: @adammil2000 Mitch's description, above, is much more clear than what's written in the following section. See my suggested change in the following section.

when Istio applied a configuration successfully. A value of `False` indicates

> there was an error.

Reviewer 2: Here it says a status value of False indicates an error, but in the following section it says that it is initialized to False, so it clearly does not always indicate an error.

Author: fixed

{{< text yaml >}}
status:
  conditions:
  - lastProbeTime: null
    lastTransitionTime: "2019-12-26T22:06:34Z"
    message: "61/122 complete"
    reason: "stillPropagating"
    status: "False"
    type: Reconciled
  - lastProbeTime: null
    lastTransitionTime: "2019-12-26T22:06:56Z"
    message: "1 Error and 1 Warning found. See validationMessages field for details"
    reason: "errorsFound"
    status: "False"
    type: PassedAnalysis
  validationMessages:
  - code: IST0101
    level: Error
    message: 'Referenced gateway not found: "bogus-gateway"'
  - code: IST0102
    level: Warn
    message: 'mTLS not enabled for virtual service'
{{< /text >}}

> ## The `type` and `status` field workflow

Reviewer 2: For consistency with the following section, this one should be titled:
```suggestion
## The `conditions` field
```

Author: fixed

The `type` field indicates the configuration condition. It can have the
following values:

* `PassedAnalysis`
* `Reconciled`

> When you apply a configuration, the `type` field immediately sets to

Reviewer 2: ```suggestion
When you apply a configuration, the `type` field is initialized to
```

Author: fixed

`Reconciled` and `status` to `False`, which blocks further commands until
`status` equals `True` or the time out period expires. The `status` field might

> transition beyond `False` instantaneously, depending on the speed of the

Reviewer 2: ```suggestion
transition to `True` instantaneously, depending on the speed of the
```

Author: fixed

cluster.

> In parallel, the `PassedAnalysis` field is set to `True` or `False` to indicate

Reviewer 2: PassedAnalysis is not a field, it's a value.

Reviewer: "The condition of type 'PassedAnalysis' has a status of 'True' or 'False' to..."

Author: fixed

the outcome of configuration validation. `PassedAnalysis` is an informational
field only, meaning that it does not block the application of an invalid
configuration. For example, it is possible for the status to indicate that
validation failed, but applying the configuration was successful. This means
Istio was able to set the new configuration, but the configuration was invalid,
likely due to a syntax error or similar problem.

The `type` field works in conjunction with the `status` field, indicating what
part of the configuration process was successful or not. For example, when the
`type` field value is `Reconciled`, it means that `status` indicates the outcome
of applying the configuration. Similarly, when `type` equals `PassedAnalysis`,
it indicates the results of validating the syntax of the CRD you provided to the
`kubectl apply` command.

## The `validationMessages` field

> In case of a validation failure, check the `validationMessages` field for

Reviewer 2: How do the multiple validation messages correspond with the conditions? 1 to 1, 2 to 2, 3 to 3, etc?

Author: @therealmitchconnors can you answer this question?

Reviewer: If the condition of type PassedValidation has a Status of 'False', there are guaranteed to be validationMessages explaining the problem.  There may be messages present even if PassedValidation Status is True, because those messages represent essentially info level messages.

Author: fixed

{{< /text >}}

In the output, the `status` field contains several nested fields with details
about the process of propagating configuration changes through the mesh. A
child field for the condition status, also named `status`, has a value of `True`
when Istio applied a configuration successfully. A value of `False` indicates
there was an error.

{{< text yaml >}}
status:
  conditions:
  - lastProbeTime: null
    lastTransitionTime: "2019-12-26T22:06:34Z"
    message: "61/122 complete"
    reason: "stillPropagating"
    status: "False"
    type: Reconciled
  - lastProbeTime: null
    lastTransitionTime: "2019-12-26T22:06:56Z"
    message: "1 Error and 1 Warning found. See validationMessages field for details"
    reason: "errorsFound"
    status: "False"
    type: PassedAnalysis
  validationMessages:
  - code: IST0101
    level: Error
    message: 'Referenced gateway not found: "bogus-gateway"'
  - code: IST0102
    level: Warn
    message: 'mTLS not enabled for virtual service'
{{< /text >}}

## The `type` and `status` field workflow

The `type` field indicates the configuration condition. It can have the
following values:

* `PassedAnalysis`
* `Reconciled`

> When you apply a configuration, the `type` field immediately sets to

Reviewer: A condition of type Reconciled and status false is added to the conditions field.  The status will become 'true' once the resource has finished reconciling.

let's remove any reference to "blocking" as that seems to be left over from kubectl wait.

Author: fixed

equals `False` is added to the `conditions` field. The status becomes `True`
when the resource finishes reconciling. The `status` field might transition to
`True` instantaneously, depending on the speed of the cluster.

The condition of type `PassedAnalysis` has a status of `True` or
`False` to indicate the outcome of configuration validation. `PassedAnalysis` is
an informational field only, meaning that it does not block the application of
an invalid configuration. For example, it is possible for the status to indicate
that validation failed, but applying the configuration was successful. This
means Istio was able to set the new configuration, but the configuration was

> invalid, likely due to a syntax error or similar problem.

Reviewer 2: 1. This section still doesn't explain what "reconciling" actually means and refers to `The condition of type ParsedAnalysis`, but doesn't explain when it was added.

2. Re: the sentence that says `PassedAnalysis is an informational field ...`, PassedAnalysis is not a field, it's a condition type.

3. This also should mention that if validation fails the reason is in the `validationMessages` field.

Any reason why you didn't accept my suggestion in https://github.com/istio/istio.io/pull/7157#discussion_r419467492?

Author: Committed via the GitHub UI but was missing after the merge fix, so I restored it.

Reviewer 2: This suggestion also seems to have gotten lost: https://github.com/istio/istio.io/pull/7157#discussion_r418208921

Reviewer 2: Also lost: https://github.com/istio/istio.io/pull/7157#discussion_r419472240

Reviewer 2: Maybe other suggestions from @therealmitchconnors are missing?

Author: Added new commits to restore what seems to be changes made via the GitHub UI.

likely due to a syntax error or similar problem.

The `type` field works in conjunction with the `status` field, indicating what
part of the configuration process was successful or not. For example, when the

> `type` field value is `Reconciled`, it means that `status` indicates the outcome

Reviewer: This paragraph should probably be removed, as it muddies the waters, and seems to imply that only one condition may exist at a time.

Author: @therealmitchconnors The selection covers two paragraphs, can you clarify which paragraph you're referring to?

Reviewer: Sorry, both paragraphs, I think.

Author: fixed

of applying the configuration. Similarly, when `type` equals `PassedAnalysis`,
it indicates the results of validating the syntax of the CRD you provided to the
`kubectl apply` command.

## The `validationMessages` field

In case of a validation failure, check the `validationMessages` field for
more information. The `validationMessages` field has details about the validation
process, such as error messages indicating that Istio cannot apply the

> configuration, and warning or informational messages that did not result in an

Reviewer: We should probably add some documentation here regarding the structure of validation messages.  @zerobfd  can you help here?

Reviewer 3: I don't think I ever got enough momentum to get the proposed spec into the API repo, so there's not really a guaranteed structure per se. We could link to https://istio.io/docs/reference/config/analysis/ here so that readers can see the various codes.

If everyone feels that not having a formal structure is insufficient here, I can see about finding time to re-start adding the structure, but I don't think it'll happen by 1.6.

Reviewer: I am working in this direction, and I expect it will have a 1.7 eta.  Collaboration welcome.  In the meantime, a short description of what is written today would be valuable.

Author: Fixed. Added a link to the config reference for now.

changes through the mesh, using the `status` field of the resource.
Status is disabled by default, and can be enabled during install with:

{{< text bash >}}
$ istioctl manifest apply --set values.pilot.env.PILOT_ENABLE_STATUS=true --set values.global.istiod.enableAnalysis=true
{{< /text >}}

The `status` field contains the state of a resource's configuration with various
informational messages, including:

* The resource's readiness.
* How many data plane instances are associated with it.
* Information for the output of tools, such as `istioctl analyze`.

For example, the `kubectl wait` command monitors the `status` field to determine
whether to unblock configuration and resume. For more information, see
[Wait for Resource Status to Apply Configuration](/docs/ops/configuration/mesh/config-resource-ready/).

## View the `status` field

To view the contents of the `status` field, use the following command:

{{< text bash >}}
$ kubectl get virtualservice <service-name> -o yaml
{{< /text >}}

In the output, the `status` field contains several nested fields with details
about the process of propagating configuration changes through the mesh.

> when Istio applied a configuration successfully.

Reviewer 2: ```suggestion
```

Author: fixed

    message: "61/122 complete"
    reason: "stillPropagating"
    status: "False"
    type: Reconciled
  - lastProbeTime: null
    lastTransitionTime: "2019-12-26T22:06:56Z"
    message: "1 Error and 1 Warning found. See validationMessages field for details"
    reason: "errorsFound"
    status: "False"
    type: PassedAnalysis
  validationMessages:
  - code: IST0101
    level: Error
    message: 'Referenced gateway not found: "bogus-gateway"'
  - code: IST0102
    level: Warn
    message: 'mTLS not enabled for virtual service'
{{< /text >}}

## The `conditions` field

Conditions represent possible states of the resource. The `type` field of a
condition can have the following values:

* `PassedAnalysis`
* `Reconciled`

When you apply a configuration, a condition of each of these types is added to the
`conditions` field.

The `status` field of the `Reconciled` type condition is initialized to `False`
to indicate the resource is still in the process of being distributed to all the proxies.
When finished reconciling, the status will become `True`. The `status` field might
transition to `True` instantaneously, depending on the speed of the cluster.

> The `status` field of the `PassedAnalysis` type condition will have a value of

Reviewer 2: ```suggestion

The `status` field of the `PassedAnalysis` type condition will have a value of
```

Author: fixed
